File: Question 0064 - Minimum Path Sum/src/Solution.java
import java.util.Arrays;

class Solution
{
    private int solve(int[][] grid, int r, int c, int[][] dp)
    {
        //base case
        if(r == grid.length - 1 && c == grid[0].length-1) return grid[r][c];
        if(r == grid.length -1) return grid[r][c] + solve(grid, r, c+1, dp);
        if(c == grid[0].length -1) return grid[r][c] + solve(grid, r+1, c, dp);
        if(dp[r][c] != -1) return dp[r][c];

        int mini;
        //right
        mini = grid[r][c] + solve(grid, r, c+1, dp);

        //down
        mini = Math.min(mini, grid[r][c] + solve(grid, r+1, c, dp));

        return dp[r][c] = mini;
    }

    public int minPathSum(int[][] grid)
    {
        int[][] dp = new int[grid.length][grid[0].length];
        Arrays.stream(dp).forEach(r->Arrays.fill(r,-1));
        return solve(grid, 0, 0, dp);
    }
}

File: Question 0064 - Minimum Path Sum/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

int solve(vector<vector<int>>& grid, int r, int c, vector<vector<int>>& dir, vector<vector<int>>& dp)
{
    if(r == grid.size() -1 && c == grid[0].size()-1) return grid[r][c];
    if(r < 0 || r >= grid.size() || c < 0 || c >= grid[0].size()) return 0;
    if(dp[r][c] != -1)  return dp[r][c];

    int mini = INT_MAX;
    for(auto& d : dir)
    {
        int x = r + d[0];
        int y = c + d[1];
        if(x >= 0 && y >= 0 && x < grid.size() && y < grid[0].size())
        {
            mini = min(mini, grid[r][c] + solve(grid, x, y, dir, dp));
        }
    }

    return dp[r][c] = mini;
}

class Solution 
{
public:
    int minPathSum(vector<vector<int>>& grid) 
    {
        vector<vector<int>> dir = {{1,0}, {0, 1}};
        vector<vector<int>> dp (grid.size(), vector<int>(grid[0].size(), -1));
        return solve(grid, 0, 0, dir, dp);
    }
};